cdap-api/src/main/java/co/cask/cdap/api/dataset/DatasetSpecification.java

```diff
@@ -134,6 +155,15 @@ public int getIntProperty(String key, int defaultValue) {
   }
 
   /**
+   * Return the original properties with which the dataset was created/reconfigured.
+   * @return an immutable map. For embedded datasets, this will always return null.
+   */
+  @Nullable
+  public Map<String, String> getOriginalProperties() {
+    return originalProperties == null ? null : Collections.unmodifiableMap(originalProperties);
```

Comment:
Maybe we can have the `originalProperties` an unmodifiable map in the constructor so that there is no new instance being created on every call to this method.